### ISSUE TITLE
`new_without_default_derive` False Positive

### ISSUE THREAD

**Issue by uHOOCCOOHu** [rust-lang-nursery/rust-clippy#2226]
It seems deriving `Default` on struct with generic type and lifetime restriction will cause it.
In this example, clippy 0.0.171 suggests me to add `#[derive(Default)]` again.
```rust
#[derive(Default)]
pub struct S<'a, T: 'a>(Option<&'a T>);

// Manually impl (remove `#[derive(Default)]`) works as expected.
// impl<'a, T: 'a> Default for S<'a, T> {
//     fn default() -> Self {
//         S::new()
//     }
// }

impl<'a, T: 'a> S<'a, T> {
    pub fn new() -> Self { 
        S(None) 
    }
}

fn main() {
    let _ = S::<i32>::default(); // works
}
```

**Comment by oli-obk** [comment]
I'm confused. What's the wrong behaviour?

**Comment by uHOOCCOOHu** [comment]
`S` derives `Debug` and it works. But clippy suggests me to add ``#[derive(Default)]`` again.

**Comment by oli-obk** [comment]
Oh... that is so weird. I'll have a look

**Comment by oli-obk** [comment]
Ok. I found the culprit:

```
error[E0277]: the trait bound `T: std::default::Default` is not satisfied
  --> tests/ui/new_without_default.rs:90:28
   |
90 |     pub fn new() -> Self { Default::default() }
   |                            ^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `T`
   |
   = help: consider adding a `where T: std::default::Default` bound
   = note: required because of the requirements on the impl of `std::default::Default` for `FalsePositive<'_, T>`
   = note: required by `std::default::Default::default`
```

`Self` really does not implement `Default`, thus we suggest to derive it when we should be suggesting to manually impl it

**Comment by uHOOCCOOHu** [comment]
The deriving impl of `Default` requires `T: Default` although we have `Option<T>: Default` for all `T`.  Is this the expected behavior? It's quite confusing.
Without lifetimes, `#[derive(Default)] pub struct S<T>(Option<T>);` also has this issue.

**Comment by oli-obk** [comment]
Yea this is a limitation of `derive`. Rustc tracks this under https://github.com/rust-lang/rust/issues/26925

**Comment by oli-obk** [comment]
Anyone interested in fixing this. Essentially you need to do the same as the `missing_debug_implementations` lint: https://github.com/rust-lang/rust/blob/2278506f682062c7988c9e534a94366c56693981/src/librustc_lint/builtin.rs#L588-L600 instead of the current logic in `can_derive_default` that calls `!implements_trait(...)`. If there are any impls of `Default` for the type, irrelevant of the generics, then just not report the lint.

A future lint could detect `derive`s that could be made more general by turning them into implementations. So essentially fixing the rust bug of bad derives by detecting bad derives and doing any of the fixes suggested in the issue.
